# Assignment 8, Part 1: Pandas

This notebook is based on the official `pandas` [documentation](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html).  Unless otherwise credited, quoted text comes from this document.

`Pandas` is a Python package that we will use to manage labeled data.

## Overview

*Where does this tool fit in my life?*

> For data scientists, working with data is typically divided into multiple stages: munging and cleaning data, analyzing / modeling it, then organizing the results of the analysis into a form suitable for plotting or tabular display. pandas is the ideal tool for all of these tasks.


*Why use this package?*

> pandas is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, real world data analysis in Python

### Data Structures

Pandas provides two new objects in which to store and manipulate data.

Dimensions | Name |	Description
-----------|-------|--------------------------------------
1 |	Series | 	1D labeled homogeneously-typed array
2 |	DataFrame | 	General 2D labeled, size-mutable tabular structure with potentially heterogeneously-typed column

Typical usage of this package involves the creation of one or more `DataFrame` objects, and manipulating them with the provided methods.

### Pandas Cheat Sheet

*"[`Package Name`] cheat sheet"* for popular packages can often be a worthwhile internet query.

In the case of pandas, there is an official cheet sheat you can find [here](http://pandas.pydata.org/Pandas_Cheat_Sheet.pdf).


### 10* Minutes to Pandas

The official [10 Minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html#min) documentation is an excellent resource, and provides the titular guide which we will drawn from and expanded upon for this tutorial.

_Note: You can access the pandas documentation from within a jupyter lab session by selecting 'help' and then selecting 'pandas' from the dropdown menu._

---
## 1. Getting Started

First, we need to import the pandas library (and Numpy library too).  All packages are imported at the top of the notebook. Execute the code in the following cell to get started with this notebook (type Ctrl+Enter in the cell below)

In [2]:
# Numpy and pandas usage are often intertwined.
# These abbreviations are ubiquitiously used.
import numpy as np
import pandas as pd

The code above imports both numpy and pandas as variables named `np` and `pd`. We can use these variables to access the functionality of both libraries respectively.  The above is what we will use for the rest of this class.

You may be wondering why we didn't import pandas like this:  
```python
import pandas
```
We could, but the first is far more commonly seen. 

#### Task 1.1: Setup


---
## 2. Data Object Creation

Here we will learn to create and use the two data structures provided by Pandas: **Series** and **DataFrames**.  For additional help, see the [Data Structure Intro section](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dsintro).


### 2.1 The `pd.Series` Data Object

In Pandas, a Series is a one dimensional array, where all the entries are of the same data type (i.e. integer, string, boolean, etc.) or **NaN**.  NaN means "not a number" but serves as a marker when a value is missing.  Before using a series take care to consider the correct data type.

The following cell creates an new series containing an array of 5 numbers and one missing value by simply passing a list of 6 elements to the `pd.Series` function. Execute the cell:

In [7]:
# Create the series object.
my_series = pd.Series([1, 3, 5, np.nan, 6, 8])

In [24]:
two_tw = pd.DataFrame({"lee": [2, 3, 5, 6, 7]})

In [25]:
type(two_tw)

pandas.core.frame.DataFrame

In [26]:
print(two_tw)

   lee
0    2
1    3
2    5
3    6
4    7


Execute the next cell to see what the new series object named `my_series` contains

In [27]:
type(my_series)

pandas.core.series.Series

In [28]:
my_series.ndim

1

In [29]:
my_series.dtype

dtype('float64')

In [30]:
# Call the object in the notebook.
# This (mostly) calls the __repr__() function of a given object.
my_series

0    1.0
1    3.0
2    5.0
3    NaN
4    6.0
5    8.0
dtype: float64

#### Task 2a Create a `pd.Series` object
Use the practice notebook to create a series of your own design.

In [112]:
series2a = pd.Series(['A', 'B', 'C', 'D'])

In [113]:
series2a.dtype

dtype('O')

In [114]:
series2a

0    A
1    B
2    C
3    D
dtype: object

### 2.2 The `pd.DataFrame` Data Object

In Pandas, a DataFrame is a two-dimensional array.  Just like Numpy, it has axes labeled "rows" and "cols".  However, unlike Numpy, it allows for different data types in each column!  DataFrames also provide greater flexibilty for indexing and slicing.

#### 2.2.1 Create a DataFrame with a dictionary

A dataframe is created by passing a Python dictionary (i.e. **dict**) to the `pd.DataFrame` function.  Remember dictionaries have key/value pairs. The dictionary is interpreted in the following way:

- The keys in the dictionary argument becomes the column names
- The value in the dictionary argument contains the column values
- Each "column" must have the same number of values.

The following cell contains code that creates a new DataFrame object named `df_1`.  This dataframe consists of two columns, one named "alpha" and the other "beta".  Each column has 5 elements. Notice that the columns contain different data types.  Execute the following cell.

In [115]:
# Creating a dataframe with a dictionary.
df_1 = pd.DataFrame(
    {'alpha': [0, 1, 2, 3, 4],
     'beta': ['a', 'b', 'c', 'd', 'e']})

Execute the next cell to see what the new data frame object named `df_1` contains

In [116]:
# Call the object in the notebook.
# This time we get a html output, as the notebook shell
# sees a _repr_html_() function to call.
df_1

alpha beta
0      0    a
1      1    b
2      2    c
3      3    d
4      4    e

In [117]:
yte = np.array([2, 3])
print(yte)

[2 3]


In [118]:
yte.dtype

dtype('int32')

In [119]:
type(yte)

numpy.ndarray

Observe that a DataFrame can be visualized as a data table or Excel spreadsheet. Also, an index, in the left-most column, has been automatically generated.

#### 2.2.2. Create a DataFrame with a Numpy array
Alternatively, we can create a DataFrame by passing in a Numpy matrix.  Remember that this is only possible if the data is all of the same type as Numpy arrays can only store one data type. The following code creates a Dataframe of 6 rows and 4 columns and fills it with random numbers generated using the Numpy `np.random.randn` function.  Execute the following cell:

In [120]:
df = pd.DataFrame(np.random.randn(6, 4))

Let's take a look at our new data frame by executing the following cell:

In [121]:
df

0         1         2         3
0 -0.217700  0.744416 -1.226374 -0.334128
1 -0.454842  0.027548  0.298928  1.170409
2 -0.326147  0.856962 -0.860298 -0.596449
3  0.820972 -1.783020 -1.709158  0.054475
4  2.108757 -1.278678 -1.404922  0.039553
5  0.035882 -0.357052 -1.332733 -1.604969

In [122]:
test = np.random.randn(6, 4)

In [123]:
test

array([[-1.58841334,  0.67386274,  0.41792552, -1.62798518],
       [ 0.2601895 ,  1.35851485, -0.75523295,  1.47696313],
       [ 1.17442729, -0.30162355, -0.13324852,  0.44083147],
       [-0.22097715,  0.73691963, -0.54326671, -0.73738865],
       [ 0.43771114, -0.97097131, -0.27990582,  0.69945903],
       [-0.69011924, -0.44949358, -0.36204499, -0.34707316]])

In [124]:
df_test = pd.DataFrame(test)

In [125]:
df_test

0         1         2         3
0 -1.588413  0.673863  0.417926 -1.627985
1  0.260190  1.358515 -0.755233  1.476963
2  1.174427 -0.301624 -0.133249  0.440831
3 -0.220977  0.736920 -0.543267 -0.737389
4  0.437711 -0.970971 -0.279906  0.699459
5 -0.690119 -0.449494 -0.362045 -0.347073

In [126]:
df_test.head(2)

0         1         2         3
0 -1.588413  0.673863  0.417926 -1.627985
1  0.260190  1.358515 -0.755233  1.476963

In [127]:
# test1 = np.random.randn(6, 4, 5)

In [128]:
# test1

In [129]:
# df_test1 = pd.DataFrame(test1) # it cannot take a 3D array, just checking

Observe that because we did not provide a dictionary, we do not have header names for each column. Rather a numeric index was automatically generated.

#### Task 2b: Create a pd.DataFrame object from a Python dictionary

### 2.3. Create a DataFrame with row and column labels

Rather than have a numeric index for each row of the DataFrame, we may want to give each row a human-readable label. This will help make the data more accessible (as will be seen later).  

To demonstrate, lets reuse the 6x4 datafame we created in section 2.2.2.  First, we need to provide our index names.  Suppose we want our row names to be dates, perhaps corresponding to when the row data was collected. The code below uses the `pd.date_range` function to do create a list of dates. If this were real data we would most likely include the dates collected as a column in the table, but for the purposes of demonstration we'll generate a series of regularly spaces dates over a 6 day period: one for each of the 6 rows:

In [130]:
dates = pd.date_range('20130101', periods=6)
dates

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06'],
              dtype='datetime64[ns]', freq='D')

Let's also suppose that we have 4 columns and we want them named 'A', 'B', 'C', and 'D'. 

We can re-create the DataFrame from section 2.2.2 with these names for the columns and the dates as row indexes by using the `index` and `columns` arguments when creating the DataFram as shown in the following cell.  Execute the cell to see the results: 

In [131]:
df = pd.DataFrame(np.random.randn(6, 4), index=dates, columns=list('ABCD'))
df

A         B         C         D
2013-01-01  0.024095  0.288909 -2.426394  0.141259
2013-01-02  0.296303  1.655626 -1.681553 -0.516838
2013-01-03 -1.419161  1.283363  0.421993  1.071561
2013-01-04  0.852828 -1.118256  1.047999 -0.729721
2013-01-05  0.625896 -1.328614  0.035641 -0.836521
2013-01-06  0.089465  0.523281  2.233991 -1.209160

#### Task 2c: Supply indexes and columns when creating a pd.DataFrame object.


---
## 3. Loading Data

Creating a DataFrame manually, as in the previous examples, would be tedious for large datasets.  Fortunately, Pandas provides a variety of functions for importing data from files.  This saves you from writing your own Python code to open a file, read in the contents and format it for inclusiong into a DataFrame. For this tutorial we will look at two of these import functions:  `pd.read_csv` and `pd.read_excel()`.

For this portion of the tutorial, we will use a rather [famous set of data concerning iris flowers](https://en.wikipedia.org/wiki/Iris_flower_data_set).  You can find a copy of the file [by clicking here]( https://gist.githubusercontent.com/curran/a08a1080b88344b0c8a7/raw/d546eaee765268bf2f487608c537c05e22e4b221/iris.csv).

![Iris flower](https://upload.wikimedia.org/wikipedia/commons/thumb/4/41/Iris_versicolor_3.jpg/193px-Iris_versicolor_3.jpg)

To use this file, create a directory named `data` in the same directory as this notebook and save the file in that directory with the name `iris.csv`.  

#### Task 3a: Download the iris.csv file.
Download the iris data and take a moment to learn about this data from the Wikipedia page cited above

### 3.1 Importing Using `pd.read_csv`

The `pd.read_csv` function can be used to read in files that are in comma-separated format or tab delimited format: two very common data formats.  This function is very flexible and has a large number of arguments. See the [`pd.read_csv` online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) for a thorough listing of arguments. 

The iris data is in CSV format and is compatible with the default arguments for the `pd.read_csv` function, therefore we can easily import this data by providing the path to the file as the only argument to the function. This will automatically create a DataFrame for us! Execute the following line to import the data:

In [132]:
iris_df = pd.read_csv('data/iris.csv')

#### Task 3b: Import the iris.csv file
In your own practice notebook, import the iris dataset.

#### Task 3c: Import a tab delimited file
Take a look at the `pd.read_csv` online documentation. Describe how you would import a tab-delimited file.

### 3.2 Importing using `pd.read_excel()`
It is common for data to be stored in Excel worksheets (provided the data is not too large).  You can easily export data from Excel to CSV or Tab delimited formats but Pandas provides  the `pd.read_excel` function to save you the time.  We will not practice loading data from Excel, but take a look at the [`read_excel` online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html) for details.

---
## 4. Explore your Data
There are a variety of functions for exploring data once you have imported it. Here we will discuss `pd.head`, `pd.tail`, `pd.sample`, and `pd.describe`.

### 4.1 The `pd.head` function

As with any Python variable, you can display its contents by simply typing the variable name and pressing the "enter" key.  Howevever, for very large data files this may not be desired.  The data used in this tutorial is not extremly large but this often not the case.

To easily examine your the first set of rows in your DataFrame, use the `head()` function of the `pd.DataFrame` object. By deafult it displays the first 5 rows. Execute the following cells to view the first five rows of the iris dataset.

In [133]:
iris_df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

This is an excellent way to check if you created your `pd.DataFrame` object correctly.  You can change the number of rows returned by providing an argument. The following requests the first 10 rows of the data frame:

In [134]:
type(iris_df)

pandas.core.frame.DataFrame

In [135]:
iris_df['species'] 

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
5         setosa
6         setosa
7         setosa
8         setosa
9         setosa
10        setosa
11        setosa
12        setosa
13        setosa
14        setosa
15        setosa
16        setosa
17        setosa
18        setosa
19        setosa
20        setosa
21        setosa
22        setosa
23        setosa
24        setosa
25        setosa
26        setosa
27        setosa
28        setosa
29        setosa
         ...    
120    virginica
121    virginica
122    virginica
123    virginica
124    virginica
125    virginica
126    virginica
127    virginica
128    virginica
129    virginica
130    virginica
131    virginica
132    virginica
133    virginica
134    virginica
135    virginica
136    virginica
137    virginica
138    virginica
139    virginica
140    virginica
141    virginica
142    virginica
143    virginica
144    virginica
145    virginica
146    virginica
147    virgini

In [136]:
iris_df.head(10)

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa
6           4.6          3.4           1.4          0.3  setosa
7           5.0          3.4           1.5          0.2  setosa
8           4.4          2.9           1.4          0.2  setosa
9           4.9          3.1           1.5          0.1  setosa

### 4.2 The `pd.tail` function
Simliar to the `pd.head` function, the `pd.tail` function lets you peak at the last 5 rows of data.  To demonstrate, execute the following cell.

In [137]:
iris_df.tail()

sepal_length  sepal_width  petal_length  petal_width    species
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

### 4.3 The `pd.sample` function
The `pd.sample` function selects a random set of rows to display. By default it only selects 1 row, we can provide the number of rows we would like (e.g. 5) by providing the number as an arugment:

In [138]:
iris_df.sample(5)

sepal_length  sepal_width  petal_length  petal_width     species
10            5.4          3.7           1.5          0.2      setosa
74            6.4          2.9           4.3          1.3  versicolor
83            6.0          2.7           5.1          1.6  versicolor
121           5.6          2.8           4.9          2.0   virginica
115           6.4          3.2           5.3          2.3   virginica

Repeat execution of the function to see a different set of randomly selected rows:

In [139]:
iris_df.sample(5)

sepal_length  sepal_width  petal_length  petal_width    species
144           6.7          3.3           5.7          2.5  virginica
22            4.6          3.6           1.0          0.2     setosa
141           6.9          3.1           5.1          2.3  virginica
42            4.4          3.2           1.3          0.2     setosa
117           7.7          3.8           6.7          2.2  virginica

### 4.4 The `pd.describe` function
The `pd.describe` function helps demonstrate the power of Pandas over Python dictionaries and Numpy arrays.  It provides a summary of statistics for the data frame in one simple function call. These statistics include the count, mean, standard deviation, min, max and quartiles.

Execute the following cell to demonstrate:

In [140]:
iris_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

#### Task 4a: Use head, tail and sample with the iris data frame.


---
## 5. Accessing Data

In this section we will learn to retreive values, slice the data frame, and use column and row labels (rather than numeric indexes) to find data. 

### 5.1 Viewing column and row labels.
Before we delve into retreiving data, lets first look at some properties of data frames for retrieving the dataframe column and row names.  This is because we often want to work with, or check the indexes and columns of a data frame using their names.  First, to retrieve the columns names we can use the `columns` property of the data frame:

In [141]:
iris_df.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

Similary, we can get a description of the index using the `index` property:

In [142]:
iris_df.index

RangeIndex(start=0, stop=150, step=1)

Observe that here our index is numeric and runs from 0 to 150, incremented by 1

#### Task 5a: Display the columns and indexes of the iris data frame.


### 5.2. Get all the values

Perhaps you would like to extract the contents of the DataFrame into a 2-dimensional Python list. You will lose the power of Pandas, but you should know how to do this. There are two ways to extract values: the new Pandas v0.24 way and the older way.  

This following code example used to be the standard way of getting the values of a dataframe as a `numpy` array:

```python
my_array = df.values
```

This is depreciated in version 0.24 of `pandas` to instead require the use of `to_numpy()`.

```python
my_array = df.to_numpy()
```

You can verify what version of `pandas` you have:

In [143]:
pd.__version__

'0.24.1'

#### Task 5b: Check the version of `pandas` you have, then convert a data frame to a numpy array.



In [144]:
# iris_df.values
iris_array = iris_df.values

In [145]:
# iris_array

In [146]:
iris_array = iris_df.to_numpy()

In [147]:
iris_array.dtype

dtype('O')

### 5.3 Basic Selections
Pandas provides a variety of row/column selectors that are similar to the way that lists and arrays are indxed in Python and Numpy. This tutorial will introduce those. However, please note:
> While standard Python / Numpy expressions for selecting and setting are intuitive and come in handy for interactive work, for production code, we recommend the optimized pandas data access methods, '.at', '.iat', '.loc' and '.iloc'.

This tutorial will also descirbe the `.at`, `.iat`, `.loc` and `.iloc` functions.

#### 5.3.1 Get a column
The easiest way to select a single column is to use the column label! When in doubt, use the `.column` property of the DataFrame to find the list of column labels.  Using the column label will result in the column being returned as a Series object.  For example, if we want the `sepal_length` column of the iris data frame we can execute the following:

In [148]:
slength = iris_df['sepal_length']
slength.head()

0    5.1
1    4.9
2    4.7
3    4.6
4    5.0
Name: sepal_length, dtype: float64

In [149]:
iris_df.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

Observe, we can use the `.head` function to get the first few rows of a Series as well as a DataFrame.

#### 5.3.2 Get rows (slicing)
Similar to slicing of Python lists, we can do the same with Pandas DataFrames:

```python
iris_df[start:end]
```

As a reminder, indexing of a list in Python uses 0-based indexing and negative numbers index in reverse order:

```
 +---+---+---+---+---+---+
 | P | y | t | h | o | n |
 +---+---+---+---+---+---+
   0   1   2   3   4   5  
  -6  -5  -4  -3  -2  -1
```

**Important:** Recall the recommendation to *not* using this type of indexing (slicing) in production with very large data. Instead use other provided functions (`.loc`, `iloc`, etc). However, to practice, let's get the first row of the iris dataset by slicing:

In [153]:
iris_df[0:0]

Empty DataFrame
Columns: [sepal_length, sepal_width, petal_length, petal_width, species]
Index: []

In [154]:
iris_df[0:1]

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa

Next lets get the first five rows:

In [155]:
iris_df[0:5] 

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

### 5.4. Selecting with `df.loc` and `df.iloc`

Aside from the column name and slicing (as shown in the previous sections) Pandas provides the `df.loc` and `df.iloc` accessor.  These accessors perform better for selecting subsets of the dataframe. The `.loc` accessor is used for slicing by columns and `.iloc` is used for slicing by rows.

### 5.4.1 Using `.loc`

The `.loc` accessor allows for the slicing of the data frame by using the labels of either rows and columns.  

The syntax is as follows:

```python
df.loc[row_start:row_end, column_start:column_end]
```

Remember, because the rows in the iris data frame are indexed using integers, we can use `.loc` to get rows using the integer values.  To get the first row:

In [156]:
iris_df.loc[0, 'species']

'setosa'

In [157]:
iris_df.loc[0]

sepal_length       5.1
sepal_width        3.5
petal_length       1.4
petal_width        0.2
species         setosa
Name: 0, dtype: object

In [158]:
pd.__version__

'0.24.1'

To get the first two rows:

In [169]:
iris_df.loc[0:1]  # Get the first two rows.

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

Observe that when only one row is returned that a Series is provided.  When multiple rows are returned then a new DataFrame object is returned.


Next, recall that we constructed a 6x4 dataframe in section 2.3 that was filled with random values and we set the row names using datetime values. The data frame was named `df`.  As a reminder here are the first 2 rows:

In [170]:
df.head(2)

A         B         C         D
2013-01-01  0.024095  0.288909 -2.426394  0.141259
2013-01-02  0.296303  1.655626 -1.681553 -0.516838

In this data frame the rows are indexed using strings that represent dates. We can therefore use those labels to slice the dataframe.  The following extracts the first row:

In [171]:
df.loc['2013-01-01']

A    0.024095
B    0.288909
C   -2.426394
D    0.141259
Name: 2013-01-01 00:00:00, dtype: float64

In [167]:
df.loc[:, 'A']

2013-01-01    0.024095
2013-01-02    0.296303
2013-01-03   -1.419161
2013-01-04    0.852828
2013-01-05    0.625896
2013-01-06    0.089465
Freq: D, Name: A, dtype: float64

Similarly we can get the first two rows:

In [172]:
df.loc['2013-01-01':'2013-01-03']

A         B         C         D
2013-01-01  0.024095  0.288909 -2.426394  0.141259
2013-01-02  0.296303  1.655626 -1.681553 -0.516838
2013-01-03 -1.419161  1.283363  0.421993  1.071561

#### 5.4.2 Using `.loc` to get columns
Remember for our iris data frame the column indexes have the names as provided by the `.columns` property.

```python
iris_df.columns
Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')
```

We can use these names to retrive values for specific columns. For example, the code below extracts rows 4 through 6 and the columns 'sepal_width' and 'petal_width':

In [100]:
iris_df.loc[4:6, 'sepal_width': 'petal_width']

sepal_width  petal_length  petal_width
4          3.6           1.4          0.2
5          3.9           1.7          0.4
6          3.4           1.4          0.3

The use of the colon in the code above between 'sepal_width' and 'petal_width' implies a range. Therefore every column between those two are included. This is why 'petal_length' appears in the output.  If we wanted to specifically select only the two columns and none other, we could provide the names in a list:

In [173]:
iris_df.loc[4:6, ['sepal_width', 'petal_width']]

sepal_width  petal_width
4          3.6          0.2
5          3.9          0.4
6          3.4          0.3

In [174]:
iris_df.loc[[4, 6], ['sepal_width', 'petal_width']]

sepal_width  petal_width
4          3.6          0.2
6          3.4          0.3

Observe that now we only have 'sepal_width', and 'petal_width' in the resulting data frame.

Moreover, we can limit the rows to only those specified in the same way:

In [105]:
iris_df.loc[[4,6], ['sepal_width', 'petal_width']]

sepal_width  petal_width
4          3.6          0.2
6          3.4          0.3

In [175]:
pwd

'C:\\Users\\linda\\hort503-2019\\Assignment08'

In [104]:
iris_df.to_csv('data/copy_iris.csv')

#### Task 5c: Make selections with `loc`

+ Select a single item with `at`.
+ Select a row slice with `loc`.
+ Select a row and column slice with `loc`.

### 5.4.2 Using `.iloc`

The `.iloc` accessor allows us to slice the data frame using an integer index, regardless of what the index (or column) labels actually are.

The syntax is the same as `.loc` with the exception of that integers are used instead of labels:

```python
df.iloc[row_start:row_end, column_start:column_end]
```
To retrieve the first row of the iris data we use the numeric index:

In [129]:
iris_df.iloc[0]  # Get the first row.

sepal_length       5.1
sepal_width        3.5
petal_length       1.4
petal_width        0.2
species         setosa
Name: 0, dtype: object

We can use any combination of integer indexes to get a subset of rows and columns.  The following examples retieves rows 2 to 5, and excludes the 'species' column.

In [130]:
iris_df.iloc[2:5, :-1] 

sepal_length  sepal_width  petal_length  petal_width
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [176]:
iris_df.loc[[2], ['sepal_length']]

sepal_length
2           4.7

In [177]:
iris_df.iloc[[2], [0]]

sepal_length
2           4.7

#### Task 5d: Make selections with `iloc` and `iat`.

+ Select a single item with `iat`.
+ Select a row slice with `iloc`.
+ Select a row and column slice with `iloc`.


### 5.5 "Fancy" indexing with `.where`

Pandas allows us to use boolean values to extract specific rows of data that meet certain conditions.To clarify this, first examine what occurs when we apply a condition to a data frame.  Recall the data frame we created in section 2.2.1.  As a reminder execute the following cell:

In [178]:
df_1

alpha beta
0      0    a
1      1    b
2      2    c
3      3    d
4      4    e

In [179]:
df_1.dtypes

alpha     int64
beta     object
dtype: object

Let's apply a condition to the data frame to find all values > 1.0

In [180]:
df_1.to_numpy() > 1.0 #~ this is a inverse

TypeError: '>' not supported between instances of 'str' and 'float'

Notice that the resulting data frame contains `True` and `False` values in place of every value that did or did not meet the condition.  We can use this new data frame to subset the original dataframe.

Lets consider the iris dataset for a more realistic example.  Suppose we want to find all rows with a `sepal_length` greater than 5.8. We know from the call to `describe` that the mean is approximately 5.8. We can do so with the following:

In [181]:
condition = iris_df['sepal_length'] > 5.8
iris_df[condition].head()

sepal_length  sepal_width  petal_length  petal_width     species
50           7.0          3.2           4.7          1.4  versicolor
51           6.4          3.2           4.5          1.5  versicolor
52           6.9          3.1           4.9          1.5  versicolor
54           6.5          2.8           4.6          1.5  versicolor
56           6.3          3.3           4.7          1.6  versicolor

In [149]:
iris_df[iris_df['sepal_length'] > 5.8]

sepal_length  sepal_width  petal_length  petal_width     species
50            7.0          3.2           4.7          1.4  versicolor
51            6.4          3.2           4.5          1.5  versicolor
52            6.9          3.1           4.9          1.5  versicolor
54            6.5          2.8           4.6          1.5  versicolor
56            6.3          3.3           4.7          1.6  versicolor
58            6.6          2.9           4.6          1.3  versicolor
61            5.9          3.0           4.2          1.5  versicolor
62            6.0          2.2           4.0          1.0  versicolor
63            6.1          2.9           4.7          1.4  versicolor
65            6.7          3.1           4.4          1.4  versicolor
68            6.2          2.2           4.5          1.5  versicolor
70            5.9          3.2           4.8          1.8  versicolor
71            6.1          2.8           4.0          1.3  versicolor
72            6.3          2.5           4.9          1.5  versicolor
73            6.1          2.8           4.7          1.2  versicolor
74            6.4          2.9           4.3          1.3  versicolor
75            6.6          3.0           4.4          1.4  versicolor
76            6.8          2.8           4.8          1.4  versicolor
77            6.7          3.0           5.0          1.7  versicolor
78            6.0          2.9           4.5          1.5  versicolor
83            6.0          2.7           5.1          1.6  versicolor
85            6.0          3.4           4.5          1.6  versicolor
86            6.7          3.1           4.7          1.5  versicolor
87            6.3          2.3           4.4          1.3  versicolor
91            6.1          3.0           4.6          1.4  versicolor
97            6.2          2.9           4.3          1.3  versicolor
100           6.3          3.3           6.0          2.5   virginica
102           7.1          3.0           5.9          2.1   virginica
103           6.3          2.9           5.6          1.8   virginica
104           6.5          3.0           5.8          2.2   virginica
..            ...          ...           ...          ...         ...
118           7.7          2.6           6.9          2.3   virginica
119           6.0          2.2           5.0          1.5   virginica
120           6.9          3.2           5.7          2.3   virginica
122           7.7          2.8           6.7          2.0   virginica
123           6.3          2.7           4.9          1.8   virginica
124           6.7          3.3           5.7          2.1   virginica
125           7.2          3.2           6.0          1.8   virginica
126           6.2          2.8           4.8          1.8   virginica
127           6.1          3.0           4.9          1.8   virginica
128           6.4          2.8           5.6          2.1   virginica
129           7.2          3.0           5.8          1.6   virginica
130           7.4          2.8           6.1          1.9   virginica
131           7.9          3.8           6.4          2.0   virginica
132           6.4          2.8           5.6          2.2   virginica
133           6.3          2.8           5.1          1.5   virginica
134           6.1          2.6           5.6          1.4   virginica
135           7.7          3.0           6.1          2.3   virginica
136           6.3          3.4           5.6          2.4   virginica
137           6.4          3.1           5.5          1.8   virginica
138           6.0          3.0           4.8          1.8   virginica
139           6.9          3.1           5.4          2.1   virginica
140           6.7          3.1           5.6          2.4   virginica
141           6.9          3.1           5.1          2.3   virginica
143           6.8          3.2           5.9          2.3   virginica
144           6.7          3.3           5.7          2.5   virginica
145           6

In [188]:
condition2 = ~(iris_df['species'] == 'versicolor')
# print(condition2)
iris_df[condition2]

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
5             5.4          3.9           1.7          0.4     setosa
6             4.6          3.4           1.4          0.3     setosa
7             5.0          3.4           1.5          0.2     setosa
8             4.4          2.9           1.4          0.2     setosa
9             4.9          3.1           1.5          0.1     setosa
10            5.4          3.7           1.5          0.2     setosa
11            4.8          3.4           1.6          0.2     setosa
12            4.8          3.0           1.4          0.1     setosa
13            4.3          3.0           1.1          0.1     setosa
14            5.8          4.0           1.2          0.2     setosa
15            5.7          4.4           1.5          0.4     setosa
16            5.4          3.9           1.3          0.4     setosa
17            5.1          3.5           1.4          0.3     setosa
18            5.7          3.8           1.7          0.3     setosa
19            5.1          3.8           1.5          0.3     setosa
20            5.4          3.4           1.7          0.2     setosa
21            5.1          3.7           1.5          0.4     setosa
22            4.6          3.6           1.0          0.2     setosa
23            5.1          3.3           1.7          0.5     setosa
24            4.8          3.4           1.9          0.2     setosa
25            5.0          3.0           1.6          0.2     setosa
26            5.0          3.4           1.6          0.4     setosa
27            5.2          3.5           1.5          0.2     setosa
28            5.2          3.4           1.4          0.2     setosa
29            4.7          3.2           1.6          0.2     setosa
..            ...          ...           ...          ...        ...
120           6.9          3.2           5.7          2.3  virginica
121           5.6          2.8           4.9          2.0  virginica
122           7.7          2.8           6.7          2.0  virginica
123           6.3          2.7           4.9          1.8  virginica
124           6.7          3.3           5.7          2.1  virginica
125           7.2          3.2           6.0          1.8  virginica
126           6.2          2.8           4.8          1.8  virginica
127           6.1          3.0           4.9          1.8  virginica
128           6.4          2.8           5.6          2.1  virginica
129           7.2          3.0           5.8          1.6  virginica
130           7.4          2.8           6.1          1.9  virginica
131           7.9          3.8           6.4          2.0  virginica
132           6.4          2.8           5.6          2.2  virginica
133           6.3          2.8           5.1          1.5  virginica
134           6.1          2.6           5.6          1.4  virginica
135           7.7          3.0           6.1          2.3  virginica
136           6.3          3.4           5.6          2.4  virginica
137           6.4          3.1           5.5          1.8  virginica
138           6.0          3.0           4.8          1.8  virginica
139           6.9          3.1           5.4          2.1  virginica
140           6.7          3.1           5.6          2.4  virginica
141           6.9          3.1           5.1          2.3  virginica
142           5.8          2.7           5.1          1.9  virginica
143           6.8          3.2           5.9          2.3  virginica
144           6.7          3.3           5.7          2.5  virginica
145           6.7          3.0           5.2          2.3  virginica
146

In [189]:
condition2 = iris_df['species'] != 'versicolor'
iris_df[condition2].head()

# print(condition2)

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

You could rewrite the two lines above to combine them into a single line with identical results:

In [190]:
iris_df[iris_df['sepal_length'] > 5.8].head()

sepal_length  sepal_width  petal_length  petal_width     species
50           7.0          3.2           4.7          1.4  versicolor
51           6.4          3.2           4.5          1.5  versicolor
52           6.9          3.1           4.9          1.5  versicolor
54           6.5          2.8           4.6          1.5  versicolor
56           6.3          3.3           4.7          1.6  versicolor

In [151]:
iris_df['sepal_length'] > 5.8

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28     False
29     False
       ...  
120     True
121    False
122     True
123     True
124     True
125     True
126     True
127     True
128     True
129     True
130     True
131     True
132     True
133     True
134     True
135     True
136     True
137     True
138     True
139     True
140     True
141     True
142    False
143     True
144     True
145     True
146     True
147     True
148     True
149     True
Name: sepal_length, Length: 150, dtype: bool

#### Task 5e: Boolean Indexing

Create subsets using boolean indexes that:
+ Use one boolean operator.
+ Use two boolean operators.
